rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -3132,10 +3279,27 @@ extern "C" rmw_ret_t rmw_send_response(
   RET_NULL(request_header);
   RET_NULL(ros_response);
   CddsService * info = static_cast<CddsService *>(service->data);
-  cdds_request_header_t header;
-  memcpy(&header.guid, request_header->writer_guid, sizeof(header.guid));
-  header.seq = request_header->sequence_number;
-  return rmw_send_response_request(&info->service, header, ros_response);
+  // Block until the response reader has been matched by the response writer (this is a
+  // workaround: rmw_service_server_is_available should keep returning false until this
+  // is a given).
```

Comment:
Isn't that possible with the addition of the new `user_data`?
Blocking doesn't sound nice

Reply:
The DDS spec doesn't say a word about what a subscription/publication match event actually guarantees and Cyclone DDS (along with most, perhaps even all, DDS implementations) treats it as a local decision, generating the event when the reader has discovered the writer. It is the most intuitive interpretation of the spec.

However, that doesn't mean it yields the most intuitive behaviour and the absence of a clear specification means there is room to make the "subscription matched" event more useful generating it only once there's proof that the writer has discovered the reader. But doing that depends in part on the behaviour of the writer implementation, and so interoperability at DDS level needs to be thought through.

I am considering doing all that in Cyclone, but there's no way I can get that released soon enough. My 0.6.0 release is already slipping because of the release process' dependency on very busy volunteers ...

Reply:
Yeah, at some point I was confused and thought that that extra information could help to solve the problem, but it can't.

I think that [this comment](https://github.com/ros2/rmw_cyclonedds/pull/187#discussion_r428660615) should be addressed, to avoid the remote chance of the wait taking forever.

I also don't quite like the idea of introducing blocking behavior, but if other people approve I'm fine with the workaround for now.